File: src/parser.js
const {UnaryOp, BinOp, Primitive, ContextValue, FunctionCall, ValueAccess, List, Object} = require("../src/AST");
const {SyntaxError} = require('./error');

let syntaxRuleError = (token, expects) => {
    expects.sort();
    return new SyntaxError(`Found ${token.value}, expected ${expects.join(', ')}`, token);
};

class Parser {
    constructor(tokenizer, source, offset = 0) {
        this._source = source;
        this._tokenizer = tokenizer;
        this.current_token = this._tokenizer.next(this._source, offset);
        this.unaryOpTokens = ["-", "+", "!"];
        this.primitivesTokens = ["number", "null", "true", "false", "string"];
        this.operations = [["||"], ["&&"], ["in"], ["==", "!="], ["<", ">", "<=", ">="], ["+", "-"], ["*", "/"], ["**"]];
        this.expectedTokens = ["!", "(", "+", "-", "[", "false", "identifier", "null", "number", "string", "true", "{"];

    }

    takeToken(...kinds) {
        if (this.current_token == null) {
            throw new SyntaxError('Unexpected end of input');
        }

        if (kinds.length > 0 && kinds.indexOf(this.current_token.kind) === -1) {
            throw syntaxRuleError(this.current_token, kinds);
        }
        try {
            this.current_token = this._tokenizer.next(this._source, this.current_token.end);
        } catch (err) {
            throw err;
        }
    }

    parse(level = 0) {
        //expr : logicalAnd (OR logicalAnd)*
        //logicalAnd : inStatement (AND inStatement)*
        //inStatement : equality (IN equality)*
        //equality : comparison (EQUALITY | INEQUALITY  comparison)*
        //comparison : addition (LESS | GREATER | LESSEQUAL | GREATEREQUAL addition)*
        //addition : multiplication (PLUS | MINUS multiplication)* "
        //multiplication : exponentiation (MUL | DIV exponentiation)*
        //exponentiation : propertyAccessOrFunc (EXP exponentiation)*
        let node;
        if (level == this.operations.length - 1) {
            node = this.parsePropertyAccessOrFunc();
            let token = this.current_token;

            for (; token != null && this.operations[level].indexOf(token.kind) !== -1; token = this.current_token) {
                this.takeToken(token.kind);
                node = new BinOp(token, this.parse(level), node);
            }
        } else {
            node = this.parse(level + 1);
            let token = this.current_token;

            for (; token != null && this.operations[level].indexOf(token.kind) !== -1; token = this.current_token) {
                this.takeToken(token.kind);
                node = new BinOp(token, node, this.parse(level + 1));
            }
        }

        return node
    }

    parsePropertyAccessOrFunc() {
        let node = this.parseUnit();
        let operators = ["[", "(", "."];
        let rightPart;
        for (let token = this.current_token; token != null && operators.indexOf(token.kind) !== -1; token = this.current_token) {
            if (token.kind == "[") {
                node = this.parseAccessWithBrackets(node)
            } else if (token.kind == ".") {
                token = this.current_token;
                this.takeToken(".");
                rightPart = new Primitive(this.current_token);
                this.takeToken("identifier");
                node = new BinOp(token, node, rightPart)
            } else if (token.kind == "(") {
                node = this.parseFunctionCall(node)
            }
        }
        return node
    }


    parseUnit() {
        // unit : unaryOp unit | primitives | contextValue | LPAREN expr RPAREN | list | object
        let token = this.current_token;
        let node;
        let isUnaryOpToken = this.unaryOpTokens.indexOf(token.kind) !== -1;
        let isPrimitivesToken = this.primitivesTokens.indexOf(token.kind) !== -1;
        if (this.current_token == null) {
            throw new SyntaxError('Unexpected end of input');
        }
        if (isUnaryOpToken) {
            this.takeToken(token.kind);
            node = new UnaryOp(token, this.parseUnit());
        } else if (isPrimitivesToken) {
            this.takeToken(token.kind);
            node = new Primitive(token);
        } else if (token.kind == "identifier") {
            this.takeToken(token.kind);
            node = new ContextValue(token);
        } else if (token.kind == "(") {
            this.takeToken("(");
            node = this.parse();
            if (node == null) {
                throw syntaxRuleError(this.current_token, this.expectedTokens);
            }
            this.takeToken(")");
        } else if (token.kind == "[") {
            node = this.parseList();
        } else if (token.kind == "{") {
            node = this.parseObject();
        }
        return node
    }

    parseFunctionCall(name) {
        //    functionCall: LPAREN (expr ( COMMA expr)*)? RPAREN
        let token = this.current_token;
        let node;
        let args = [];
        this.takeToken("(");

        if (this.current_token.kind != ")") {
            node = this.parse();
            args.push(node);

            while (this.current_token != null && this.current_token.kind == ",") {
                if (args[args.length - 1] == null) {
                    throw syntaxRuleError(this.current_token, this.expectedTokens);
                }
                this.takeToken(",");
                node = this.parse();
                args.push(node)
            }
        }
        this.takeToken(")");

        node = new FunctionCall(token, name, args);

        return node
    }

    parseList() {
        //    list : LSQAREBRAKET (expr ( COMMA expr)*)? RSQAREBRAKET)
        let node;
        let arr = [];
        let token = this.current_token;
        this.takeToken("[");

        if (this.current_token.kind != "]") {
            node = this.parse();
            arr.push(node);

            while (this.current_token.kind == ",") {
                if (arr[arr.length - 1] == null) {
                    throw syntaxRuleError(this.current_token, this.expectedTokens);
                }
                this.takeToken(",");
                node = this.parse();
                arr.push(node)
            }
        }
        this.takeToken("]");
        node = new List(token, arr);
        return node
    }

    parseAccessWithBrackets(node) {
        //    valueAccess : LSQAREBRAKET expr |(expr? SEMI expr?)  RSQAREBRAKET)
        let leftArg = null, rightArg = null;
        let token = this.current_token;
        let isInterval = false;

        this.takeToken("[");
        if (this.current_token.kind == "]") {
            throw syntaxRuleError(this.current_token, this.expectedTokens);
        }

        if (this.current_token.kind != ":") {
            leftArg = this.parse();
        }
        if (this.current_token.kind == ":") {
            isInterval = true;
            this.takeToken(":");
        }
        if (this.current_token.kind != "]") {
            rightArg = this.parse();
        }

        if (isInterval && rightArg == null && this.current_token.kind != "]") {
            throw syntaxRuleError(this.current_token, this.expectedTokens);
        }
        this.takeToken("]");
        node = new ValueAccess(token, node, isInterval, leftArg, rightArg);

        return node;
    }

    parseObject() {
        //    object : LCURLYBRACE ( STR | ID SEMI expr (COMMA STR | ID SEMI expr)*)? RCURLYBRACE (DOT ID)?
        let node;
        let obj = {};
        let key, value;
        let objToken = this.current_token;
        this.takeToken("{");
        let token = this.current_token;

        while (token != null && (token.kind == "string" || token.kind == "identifier")) {
            key = token.value;
            if (token.kind == "string") {
                key = parseString(key);
            }
            this.takeToken(token.kind);
            this.takeToken(":");
            value = this.parse();
            if (value == null) {
                throw syntaxRuleError(this.current_token, this.expectedTokens);
            }
            obj[key] = value;
            if (this.current_token != null && this.current_token.kind == "}") {
                break;
            } else {
                this.takeToken(",")
            }
            token = this.current_token;
        }
        this.takeToken("}");
        node = new Object(objToken, obj);

        return node;
    }

}

let
    parseString = (str) => {
        return str.slice(1, -1);
    };


exports
    .Parser = Parser;

File: internal/interpreter/parser/parser.go
package parser

import (
	"fmt"
)

type Parser struct {
	source              string
	tokenizer           Tokenizer
	CurrentToken        Token
	unaryOpTokens       []string
	primitivesTokens    []string
	operatorsByPriority [][]string
	expectedTokens      []string
}

func (p *Parser) NewParser(source string, tokenizer Tokenizer, offset int) (err error) {
	p.source = source
	p.tokenizer = tokenizer
	p.CurrentToken, err = p.tokenizer.Next(p.source, offset)
	p.unaryOpTokens = []string{"-", "+", "!"}
	p.primitivesTokens = []string{"number", "null", "true", "false", "string"}
	p.operatorsByPriority = [][]string{{"||"}, {"&&"}, {"in"}, {"==", "!="}, {">", "<", "<=", ">="}, {"+", "-"}, {"*", "/"}, {"**"}}
	p.expectedTokens = []string{"!", "(", "+", "-", "[", "false", "identifier", "null", "number", "string", "true", "{"}
	return
}

func (p *Parser) takeToken(kinds ...string) error {
	var err error
	if p.CurrentToken.IsEmpty() {
		return SyntaxError{
			Message: "unexpected end of input",
			Source:  p.source,
			Start:   len(p.source) - 1,
			End:     len(p.source),
		}
	}
	if len(kinds) > 0 && !StringsContains(p.CurrentToken.Kind, kinds) {
		return SyntaxError{
			Message:  fmt.Sprintf("unexpected '%s'", p.CurrentToken.Value),
			Source:   p.source,
			Start:    p.CurrentToken.Start,
			End:      p.CurrentToken.End,
			Expected: kinds,
		}
	}
	p.CurrentToken, err = p.tokenizer.Next(p.source, p.CurrentToken.End)
	return err
}

func (p *Parser) Parse(level int) (node IASTNode, err error) {
	//expr : logicalAnd (OR logicalAnd)*
	//logicalAnd : inStatement (AND inStatement)*
	//inStatement : equality (IN equality)*
	//equality : comparison (EQUALITY | INEQUALITY  comparison)*
	//comparison : addition (LESS | GREATER | LESSEQUAL | GREATEREQUAL addition)*
	//addition : multiplication (PLUS | MINUS multiplication)* "
	//multiplication : exponentiation (MUL | DIV exponentiation)*
	//exponentiation : propertyAccessOrFunc (EXP exponentiation)*
	var binaryNode BinOp
	var next IASTNode
	if level == len(p.operatorsByPriority)-1 {
		node, err = p.parsePropertyAccessOrFunc()
		if err != nil {
			return nil, err
		}
		token := p.CurrentToken

		for ; token != (Token{}) && StringsContains(token.Kind, p.operatorsByPriority[level]); token = p.CurrentToken {
			err = p.takeToken(token.Kind)
			if err != nil {
				return nil, err
			}
			next, err = p.Parse(level)
			if err != nil {
				return nil, err
			}
			binaryNode.NewNode(token, next, node)
			node = binaryNode
		}
	} else {
		node, err = p.Parse(level + 1)
		if err != nil {
			return nil, err
		}
		token := p.CurrentToken

		for ; token != (Token{}) && StringsContains(token.Kind, p.operatorsByPriority[level]); token = p.CurrentToken {
			err = p.takeToken(token.Kind)
			if err != nil {
				return nil, err
			}
			next, err = p.Parse(level + 1)
			if err != nil {
				return nil, err
			}
			binaryNode.NewNode(token, node, next)
			node = binaryNode
		}
	}

	return
}

func (p *Parser) parsePropertyAccessOrFunc() (node IASTNode, err error) {
	//propertyAccessOrFunc : unit (accessWithBrackets | DOT id | functionCall)*
	var primitiveNode ASTNode
	var binaryNode BinOp
	node, err = p.parseUnit()
	if err != nil {
		return nil, err
	}
	operators := []string{"[", "(", "."}

	for token := p.CurrentToken; p.CurrentToken != (Token{}) && StringsContains(token.Kind, operators); token = p.CurrentToken {
		switch token.Kind {
		case "[":
			node, err = p.parseAccessWithBrackets(node)
			if err != nil {
				return nil, err
			}
		case ".":
			token = p.CurrentToken
			err = p.takeToken(".")
			if err != nil {
				return nil, err
			}
			primitiveNode.NewNode(p.CurrentToken)
			rightPart := primitiveNode
			err = p.takeToken("identifier")
			if err != nil {
				return nil, err
			}
			binaryNode.NewNode(token, node, rightPart)
			node = binaryNode
		case "(":
			node, err = p.parseFunctionCall(node)
			if err != nil {
				return nil, err
			}
		}
	}

	return
}

func (p *Parser) parseUnit() (node IASTNode, err error) {
	// unit : unaryOp unit | primitives | contextValue | LPAREN expr RPAREN | list | object
	var unaryNode UnaryOp
	var primitiveNode ASTNode
	var contextValueNode ContextValue
	var next IASTNode
	if p.CurrentToken.IsEmpty() {
		return nil, SyntaxError{
			Message: "unexpected end of input",
			Source:  p.source,
			Start:   len(p.source) - 1,
			End:     len(p.source),
		}
	}
	token := p.CurrentToken
	isUnaryOpToken := StringsContains(token.Kind, p.unaryOpTokens)
	isPrimitivesToken := StringsContains(token.Kind, p.primitivesTokens)

	if isUnaryOpToken {
		err = p.takeToken(token.Kind)
		if err != nil {
			return nil, err
		}
		next, err = p.parseUnit()
		if err != nil {
			return nil, err
		}
		unaryNode.NewNode(token, next)
		node = unaryNode
	} else if isPrimitivesToken {
		err = p.takeToken(token.Kind)
		if err != nil {
			return nil, err
		}
		primitiveNode.NewNode(token)
		node = primitiveNode
	} else if token.Kind == "identifier" {
		err = p.takeToken(token.Kind)
		if err != nil {
			return nil, err
		}
		contextValueNode.NewNode(token)
		node = contextValueNode
	} else if token.Kind == "(" {
		err = p.takeToken("(")
		if err != nil {
			return nil, err
		}
		node, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		if node == nil {
			return nil, SyntaxError{
				Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
				Source:   p.source,
				Start:    p.CurrentToken.Start,
				End:      p.CurrentToken.End,
				Expected: p.expectedTokens,
			}
		}
		err = p.takeToken(")")
		if err != nil {
			return nil, err
		}
	} else if token.Kind == "[" {
		node, err = p.parseList()
		if err != nil {
			return nil, err
		}
	} else if token.Kind == "{" {
		node, err = p.parseObject()
		if err != nil {
			return nil, err
		}
	}
	return
}

func (p *Parser) parseFunctionCall(name IASTNode) (node IASTNode, err error) {
	//    functionCall: LPAREN (expr ( COMMA expr)*)? RPAREN
	var args []IASTNode
	var functionCallNode FunctionCall
	token := p.CurrentToken
	err = p.takeToken("(")
	if err != nil {
		return nil, err
	}

	if p.CurrentToken.Kind != ")" {
		node, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		args = append(args, node)

		for p.CurrentToken != (Token{}) && p.CurrentToken.Kind == "," {
			if args[len(args)-1] == nil {
				return nil, SyntaxError{
					Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
					Source:   p.source,
					Start:    p.CurrentToken.Start,
					End:      p.CurrentToken.End,
					Expected: p.expectedTokens,
				}
			}
			err = p.takeToken(",")
			if err != nil {
				return nil, err
			}
			node, err = p.Parse(0)
			if err != nil {
				return nil, err
			}
			args = append(args, node)
		}
	}
	err = p.takeToken(")")
	if err != nil {
		return nil, err
	}
	functionCallNode.NewNode(token, name, args)
	node = functionCallNode

	return
}

func (p *Parser) parseList() (node IASTNode, err error) {
	//    list : LSQAREBRAKET (expr ( COMMA expr)*)? RSQAREBRAKET)
	var list []IASTNode
	var listNode List
	var token = p.CurrentToken
	err = p.takeToken("[")
	if err != nil {
		return nil, err
	}

	if p.CurrentToken.Kind != "]" {
		node, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		list = append(list, node)

		for p.CurrentToken.Kind == "," {
			if list[len(list)-1] == nil {
				return nil, SyntaxError{
					Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
					Source:   p.source,
					Start:    p.CurrentToken.Start,
					End:      p.CurrentToken.End,
					Expected: p.expectedTokens,
				}
			}
			err = p.takeToken(",")
			if err != nil {
				return nil, err
			}
			node, err = p.Parse(0)
			if err != nil {
				return nil, err
			}
			if node == nil {
				return nil, SyntaxError{
					Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
					Source:   p.source,
					Start:    p.CurrentToken.Start,
					End:      p.CurrentToken.End,
					Expected: p.expectedTokens,
				}
			}
			list = append(list, node)
		}
	}
	err = p.takeToken("]")
	if err != nil {
		return nil, err
	}
	listNode.NewNode(token, list)
	node = listNode
	return
}

func (p *Parser) parseAccessWithBrackets(node IASTNode) (IASTNode, error) {
	//    valueAccess : LSQAREBRAKET expr |(expr? SEMI expr?)  RSQAREBRAKET)
	var arrayNode ValueAccess
	var left, right IASTNode
	var err error
	token := p.CurrentToken
	isInterval := false

	err = p.takeToken("[")
	if err != nil {
		return nil, err
	}
	if p.CurrentToken.Kind == "]" {
		return nil, SyntaxError{
			Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
			Source:   p.source,
			Start:    p.CurrentToken.Start,
			End:      p.CurrentToken.End,
			Expected: p.expectedTokens,
		}
	}
	if p.CurrentToken.Kind != ":" {
		left, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
	}
	if p.CurrentToken.Kind == ":" {
		isInterval = true
		err = p.takeToken(":")
		if err != nil {
			return nil, err
		}
	}
	if p.CurrentToken.Kind != "]" {
		right, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
	}
	if isInterval && right == nil && p.CurrentToken.Kind != "]" {
		return nil, SyntaxError{
			Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
			Source:   p.source,
			Start:    p.CurrentToken.Start,
			End:      p.CurrentToken.End,
			Expected: p.expectedTokens,
		}
	}

	err = p.takeToken("]")
	if err != nil {
		return nil, err
	}
	arrayNode.NewNode(token, node, isInterval, left, right)
	node = arrayNode

	return node, err
}

func (p *Parser) parseObject() (node IASTNode, err error) {
	//    object : LCURLYBRACE ( STR | ID COLON expr (COMMA STR | ID COLON expr)*)? RCURLYBRACE
	var objectNode Object
	var objValue IASTNode
	obj := make(map[string]IASTNode)
	token := p.CurrentToken
	err = p.takeToken("{")
	if err != nil {
		return nil, err
	}

	for p.CurrentToken.Kind == "string" || p.CurrentToken.Kind == "identifier" {
		key := p.CurrentToken.Value
		if p.CurrentToken.Kind == "string" {
			key = parseString(key)
		}
		err = p.takeToken(p.CurrentToken.Kind)
		if err != nil {
			return nil, err
		}
		err = p.takeToken(":")
		if err != nil {
			return nil, err
		}
		objValue, err = p.Parse(0)
		if objValue == nil {
			return nil, SyntaxError{
				Message:  fmt.Sprintf("Found '%s'", p.CurrentToken.Kind),
				Source:   p.source,
				Start:    p.CurrentToken.Start,
				End:      p.CurrentToken.End,
				Expected: p.expectedTokens,
			}
		}
		if err != nil {
			return nil, err
		}
		obj[key] = objValue
		if p.CurrentToken.Kind == "}" {
			break
		} else {
			err = p.takeToken(",")
			if err != nil {
				return nil, err
			}
		}
	}
	err = p.takeToken("}")
	if err != nil {
		return nil, err
	}
	objectNode.NewNode(token, obj)
	node = objectNode

	return
}

func parseString(s string) string {
	return s[1 : len(s)-1]
}
